src/poseidon.rs

```diff
@@ -0,0 +1,141 @@
+use crate::{Spec, State};
+use pairing::arithmetic::FieldExt;
+
+/// Poseidon hasher that maintains state and inputs and yields single element
+/// output when desired
+#[derive(Debug, Clone)]
+pub struct Poseidon<F: FieldExt, const T: usize, const RATE: usize> {
+    state: State<F, T>,
+    spec: Spec<F, T, RATE>,
+    absorbing: Vec<F>,
+}
+
+impl<F: FieldExt, const T: usize, const RATE: usize> Poseidon<F, T, RATE> {
+    /// Constructs a clear state poseidon instance
+    pub fn new(r_f: usize, r_p: usize) -> Self {
+        Self {
+            spec: Spec::new(r_f, r_p),
+            state: State::default(),
+            absorbing: Vec::new(),
+        }
+    }
+
+    /// Appends elements to the absorbation line
+    pub fn update(&mut self, elements: &[F]) {
+        // TODO: if `RATE` is full spawn a permutation here
+        self.absorbing.extend_from_slice(elements);
+    }
+
+    /// Results a single element by absorbing already added inputs
+    pub fn squeeze(&mut self) -> F {
+        let mut padding_offset = 0;
+        let input_elements = self.absorbing.clone();
+        for chunk in input_elements.chunks(RATE) {
+            // `padding_offset` is expected to be in [0, RATE-1]
+            padding_offset = RATE - chunk.len();
+            let mut chunk = chunk.to_vec();
+            if padding_offset > 0 {
+                chunk.push(F::one())
+            }
+
+            // Add new chunk of inputs for the next permutation cycle
+            for (input_element, state) in chunk.iter().zip(self.state.0.iter_mut().skip(1)) {
+                state.add_assign(input_element);
+            }
+            self.permute();
+        }
+
+        // TODO/FIX: can we avoid this in the on going transcript squeezing context?
```

Comment:
I don't think so. 

I could be wrong OFC but my understanding was that you need to apply padding for each squeeze you perform.

Otherways, assuming for example the padding rule of keccak:
```
Imagine:

We want to hash: [33, 15, 42].

Padded is: [33, 15, 42, 0x80, 0, 0x01]

And then we want to hash: [33, 15, 42, 0x80, 0, 0x01] directly.

These two would collide
```

Reply:
Would collide because we need to padd also the second entry for example even it has the  full pad-width.

Reply:
I was imagining in our context inputs which are evals and commitments would be more random rather than 0x01 and padding magics like 0x01. But OFC I think it can be an insecure approach.

Reply:
I'm not an expert on the security on these things. 

And agree the probability is low. 
But anyway, we should be careful with these optimizations. 
We can check with Dmitry and see whether this is a possibility.

Reply:
Also agree that we should apply what book says for now.